### PR TITLE
schemachanger: consider column target status after resolving

### LINKED ISSUE #120226: sql: ALTER TABLE ... DROP COLUMN x, ADD COLUMN x does not work correctly
**Describe the problem**

Removing a column and adding it back in the same statement does not work.

**To Reproduce**

```
CREATE TABLE "Dog" (
    id              SERIAL PRIMARY KEY,
    name            TEXT NOT NULL,
    is_good_dog     BOOLEAN NOT NULL DEFAULT TRUE
);

set sql_safe_updates=false;

ALTER TABLE "Dog" DROP COLUMN is_good_dog, ADD COLUMN is_good_dog INTEGER NOT NULL DEFAULT 100;
```

results in 
```
ERROR: column "is_good_dog" of relation "Dog" already exists
SQLSTATE: 42701
```

**Expected behavior**
Should work without error.

The problem might be that the lookup for the column doesn't check if the column is dropped.

Jira issue: CRDB-36556

### PR DESCRIPTION
**Pull request by annrpom**
### schemachanger: consider "dropping" columns during drop col

Previously, using the "comma syntax" for dropping a col
with the same name (`alter table t drop column i, drop column i;`)
led to an internal error. This patch makes it so that
the error will instead be a proper external error by
ensuring that resolving columns in the DSC considers the
"dropping" state.

Epic: none

Release note: None

---
### schemachanger: consider "dropping" columns while adding col

This patch enables the "comma syntax" to drop and add
a column with the same name in our DSC.

Epic: none
Fixes: https://github.com/cockroachdb/cockroach/issues/120226

Release note (bug fix): Dropping and adding a column with the same
name no longer results in a "column already exists error".

### CODE REVIEW

**Review by fqazi — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122389)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/122389#-Nvb4AAj2uKPtCuzQc6B:-Nvb4AAj2uKPtCuzQc6C:b-21r846) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41198de6f3112be22867ff2abbe2f15a6cc3547/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L162)):*
> ```Go
> 
> 	if col != nil && colTargetStatus == scpb.ToAbsent {
> 		return nil, nil, true
> ```

I think the side effect of the second drop should be not being able to find the column? More like a normal external error. Also I wonder if the resolve should have surfaced this :thinking:

**Review by annrpom — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122389)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/122389#-Nvb4AAj2uKPtCuzQc6B:-NvbIXnrAoIzj62BOncj:bnsk5t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41198de6f3112be22867ff2abbe2f15a6cc3547/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L162)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think the side effect of the second drop should be not being able to find the column? More like a normal external error. Also I wonder if the resolve should have surfaced this :thinking:
</blockquote></details>

yeah; that's likely better — the no-op was inspired by trying to follow what the legacy schema changer does with:
```
BEGIN;
ALTER TABLE t DROP COLUMN i;
ALTER TABLE t DROP COLUMN i;
COMMIT;
```

let me fix this

edit: er -- what the legacy schemachanger does for comma syntax as well

**Comment by annrpom** [comment]
>  Also I wonder if the resolve should have surfaced this 🤔

i think we would have to ensure that `resolve` has another param that represents the target status of the column we are wanting to transition to (target-target status?) so we can distinguish between a dropping column we don't care about vs one we do:

```
create table t (i int);

alter table t drop column i, drop column i;

> ERROR
```

vs

```
create table t (i int);

alter table t drop column i, add column i bool;

> SUCCESS
```

the other alternative is what i did. i did it because it's a simpler code change, but how do we feel about this? if we like this, i can add an `N.B.` for `resolve` that calls out the fact that column statuses aren't considered in its logic

**Review by fqazi — APPROVED** [review]
That sounds okay, and longer term we can potentially revisit later, since that looks far riskier :lgtm: 

> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122389#-:-Nvd1FgLBH-Mmr7vFrtj:byk385m)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/122389#-Nvb4AAj2uKPtCuzQc6B:-Nvd0upC7IFVwWtfkOnk:b-ui59wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d41198de6f3112be22867ff2abbe2f15a6cc3547/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L162)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

yeah; that's likely better — the no-op was inspired by trying to follow what the legacy schema changer does with:
```
BEGIN;
ALTER TABLE t DROP COLUMN i;
ALTER TABLE t DROP COLUMN i;
COMMIT;
```

let me fix this
</blockquote></details>

Aah, I think blocking the same drop is probably saner in the long term.

**Comment by annrpom** [comment]
TFTR! ('-')7

bors r=fqazi

**Comment by rafiss** [comment]
re: "This patch enables the "comma syntax" to drop and add a column with the same name in our DSC." -- I don't see that. Did something get missed?

Also, I think we should backport this to 24.1, since the comma syntax is getting enabled.

File: pkg/sql/logictest/testdata/logic_test/alter_table [hunk 1]
```diff
@@ -3841,3 +3841,18 @@ CREATE TABLE t_identity_drop_dependency(id INT PRIMARY KEY DEFAULT nextval('publ
...
 ALTER TABLE t_identity_drop ALTER COLUMN b DROP IDENTITY;
+
+subtest comma_syntax
+
+statement ok
+CREATE TABLE foo (i int);
+
+skipif config local-legacy-schema-changer
+statement ok
+ALTER TABLE foo DROP COLUMN i, ADD COLUMN i bool;
```
Comment (annrpom) [inline pkg/sql/logictest/testdata/logic_test/alter_table:3852]: i have a test here @rafiss

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go [hunk 2]
```diff
@@ -69,7 +69,9 @@ func alterTableAddColumn(
 					"column name %q conflicts with a system column name",
 					d.Name))
 			}
-			panic(sqlerrors.NewColumnAlreadyExistsInRelationError(string(d.Name), tn.Object()))
+			if colTargetStatus != scpb.ToAbsent {
```
Comment (annrpom) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:72]: and the code change @rafiss 

(unless i worded something incorrectly for the release note)

File: pkg/sql/logictest/testdata/logic_test/alter_table (hunk 1 above)
Comment (rafiss) [reply pkg/sql/logictest/testdata/logic_test/alter_table:3852]: oops - i misread and thought the commit said the testing was added to the RSW.

**Comment by annrpom** [comment]
blathers backport 24.1

**Pull request merged**